### ISSUE TITLE
better way to init

### ISSUE THREAD

**Issue by mastef** [tyxla/gravity-forms-section-tabs#12]
It might be better to inline the init javascript in `gform_register_init_scripts` like in this example : https://www.gravityhelp.com/documentation/article/gform_register_init_scripts/

The use-case where the current behaviour is breaking is when multiple forms are loaded on a page through ajax, and the initial 'gform_post_render' binding is being cleaned up between form loads with `jQuery(document).unbind('gform_post_render');`. ( This has been so far necessary so the input character counter doesn't get loaded multiple times on each consecutive form... )

So currently the section tabs then don't get initialized for the 2nd ajax form that was loaded, because the event listener on gform_post_render was removed, and then re-bound on the next ajax form load. Adding the listener with `gform_register_init_scripts` would add them after each ajax form, so that would keep the init on the safe side. Could be something like this :

``` php
add_action( 'gform_register_init_scripts', 'gform_section_tabs' );
function gform_section_tabs( $form ) {

    $script = "(function($){
        $('form.gravity_forms_section_tabs_enabled').each(function() {                                                                                                                                                                            
                initSectionTabs($(this).attr('id').replace('gform_', ''));                                                                                                                                                                        
        });
    })(jQuery);";

    GFFormDisplay::add_init_script( $form['id'], 'section_tabs', GFFormDisplay::ON_PAGE_RENDER, $script );
}
```

( This example would still not work, since `initSectionTabs` in the main.js is not in the public scope, so it would have to be moved maybe to e.g. window.gravity_forms_section_tabs.initSectionTabs to be able to call it from other functions )

What do you think?

**Comment by tyxla** [comment]
Hi @mastef! 

Thanks for the great description of your issue. 

AFAIK, the `gform_post_render` event should be called each time after the form has been rendered. If it is unbound, it has to be bound again because there are many plugins and scripts that depend on it. Can you please clarify where exactly it gets unbound from the `document`?

Also, I'm personally against inline scripts and I consider them bad code practice. But that's in general - sometimes there are no other ways and we have to go with that. The JS of Gravity Forms is very hacky and we often have to deal with it in the same manner.

So, I'd love to see some more details from you in terms of how I can reproduce this issue. Once I can reproduce the issue, I will consider greating a workaround in the plugin for fixing it. 
So basically, I need step-by-step instructions (and probably essentials of your code) in order to reproduce the issue.

Thanks in advance.

**Comment by mastef** [comment]
Hey @tyxla,

So, the way it happens now is, when you load a gravity forms over ajax, it binds the events it needs to `gform_post_render` parts after the form, based on the `gform_register_init_scripts` action. That means if a form needs the max-character counter, it will attach it to `gform_post_render` after it was loaded. The problem appears when you want to close the ajax form, and open it again ( or open another one ).

So the example is based on embedding a form in a bootstrap modal. The bootstrap modal simply opens a remote url with the GF function call based on the form id ( [Embedding gf form](https://www.gravityhelp.com/documentation/article/embedding-a-form/#function-call) ). The container page has all necessary scripts embedded with `gravity_form_enqueue_scripts($id, true);`.

The reproducible test is quite simple : 
1. create a gf form with an input field that is character limited
2. open a bootstrap modal with an ajax gf form
3. close it
4. open it again

Result : The input field will have 2x the '5 out of 20 characters' part attached, because `gform_post_render` was executed with twice the events - step 4 bound the same functions _again_ to `gform_post_render` after the 2nd instance of the form was loaded. And as many times as you close and re-open it, as many input field counters will be attached.

![issue](http://new.tinygrab.com/6c91891dfe012fcfbe2500103406d46c3385c09afe.png)

So every time a new form is loaded, it attaches all events to `gform_post_render` - and not uniquely. It attaches them again and again.

Now the way to solve it - at least from what I've discovered so far : 
1. create a gf form with an input field that is character limited
2. open a boostrap modal with an ajax gf form ( remember to `gravity_form_enqueue_scripts` )
3. close it
4. clean up e.g. on 'hidden.bs.modal' by unbinding `gform_post_render`, `gform_post_conditional_logic`, etc
5. open it again

And that works then as expected, as all `gform_post_render` hooks are only attached and fired once. But it doesn't work with `gravity-forms-section-tabs` as it only attached to `gform_post_render` on page load. So the 2nd form doesn't have the right events to fire to enable it.

I fully understand your aversion to inline scripting, and I'm also on the same page; but we kind of have to abide by the plugin architecture unfortunately. If there'd be another way that would be great, but I haven't discovered e.g. another event firing on form load before `gform_post_render` that could be listened to, which could then trigger a bind to `gform_post_render` itself... it seems as if the inline script solution through `gform_register_init_scripts` is the key currently.

So in terms of plugin scripting we have to expect that `gform_post_render` will be only bound when the form is loaded, and not on page load, and the way to attach it then is through `gform_register_init_scripts`.

We wouldn't have this problem if events could be bound easily only once - e.g. duplicate .bind() being prevented. But from what I could see this doesn't happen currently, and there is no 'easy' way with jquery at the moment.

**Comment by mastef** [comment]
Hold on, unbinding it might stop the multifile uploader from working as well. Let me dig a bit more into it... all in all it's a bit messy with different methods being implemented, and some of them are of course not targeted at ajax...

Maybe this is just a misbehaviour of the text counter plugin. But anyway troublesome that each new form load in a modal/ajax will attach all kinds of functions on top of the old ones. There should be a cleaner way

**Comment by mastef** [comment]
For now I got around by hacking the counter script, so it doesn't re-initialize if it attached itself already... nothing to do with your plugin, but just for reference in case anybody looks for a solution to this...

```
add_filter( 'gform_counter_script', 'set_counter_script', 10, 4 );

function set_counter_script( $script, $form_id, $field_id, $max_length ) {
    $script = "

    if( !jQuery('#{$field_id}').next().is('.charleft')){
    jQuery('#{$field_id}').textareaCount(" .
                "    {" .
                "    'maxCharacterSize': {$max_length}," .
                "    'originalStyle': 'ginput_counter'," .
                "    'displayFormat' : '#input / #max'" .
                "    });" . 
    " } ";
    return $script;
}
```

Back to topic : I guess if GF would load the post-form js in a jQuery namespace bind this could solve this.. `jQuery(document).unbind('gform_post_render.form1').bind('gform_post_render.form1', function({ ... }));`. This way it would unload previous form-related post render binds, and re-bind them while leaving other globally loaded binds alone. I'll suggest it to the GF guys

**Comment by tyxla** [comment]
Makes sense to me. Thanks!
